File: BST.rb
# Actual Tree for Testing

class Node
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

root = Node.new(5)
root.left = Node.new(2)
root.left.left = Node.new(1)
root.left.right = Node.new(4)
root.left.right.left = Node.new(3)

root.right = Node.new(7)
root.right.left = Node.new(6)
root.right.right = Node.new(9)
root.right.right.left = Node.new(8)
root.right.right.right = Node.new(10)

#           5
#       /       \
#     2           7
#   /   \       /   \
# 1       4    6     9
#       /          /   \
#     3           8     10


# given a node in BST, find next largest value

def next_largest(node)
  if node.right
    return node.right.val if node.right.left.nil?
    child = node.right
    while true
      return child.left.val if child.left.left.nil?
      child = child.left
    end
  end

  if node.parent.left == node
    return node.parent.val
  end

  if node.parent.right == node
    parent = node.parent
    while true
      return nil if parent.parent.nil?
      if parent.parent.left == parent
        return parent.parent.val
      end
      parent = parent.parent
    end
  end

end

# find max depth of binary tree
def max_depth(root)
  return 0 if root.nil?

  if root.left || root.right
    depth = 1
  end

  return depth + (max_depth(root.left), max_depth(root.right)).max

end

File: 2020/BST.js
//=============================================================================================
// PAY ATTENTION TO BIG O HERE

// Count BST nodes that lie in a given range
// Given a Binary Search Tree (BST) and a range, count number of nodes that lie in the given range.
// Examples:

// Input:
//         10
//       /    \
//     5       50
//    /       /  \
//  1       40   100
//                  \
//                  101
// Range: [5, 45]

// Output:  3
// There are three nodes in range, 5, 10 and 40

// O(k + h) k is number of nodes in range + h is height of tree since we need to check along 2 boundries all the way down for potential child nodes in range
const countBSTNodesInGivenRange = function(root, range) {
    let count = 0;
    const recurse = function (node) {
        if (!node) return;

        if (node.val >= range[0] && node.val <= range[1]) count++;
        if (node.val >= range[0]) recurse(node.left);
        if (node.val <= range[1]) recurse(node.right);
    }
    recurse(root);
    return count;
}
// console.log(countBSTNodesInGivenRange({
//     val: 10,
//     left: { val: 5, left: { val: 1 } },
//     right: { val: 50, left: { val: 40 }, right: { val: 100, right: { val: 101 } } }
// }, [5,45]))

//=============================================================================================
